run-python/python.go

```diff
@@ -0,0 +1,266 @@
+package runner
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/stoic-cli/stoic-cli-core"
+)
+
+const (
+	getPipCacheKey = "run-python/get-pip.py"
+	getPipURL      = "https://bootstrap.pypa.io/get-pip.py"
+
+	pipCacheBase     = ".pip-cache"
+	readyBase        = ".ready"
+	requirementsBase = ".requirements"
+
+	pipRequirements = "" +
+		"pip==10.0.1 --hash=sha256:717cdffb2833be8409433a93746744b59505f42146e8d37de6c62b430e25d6d7\n" +
+		"setuptools==39.1.0 --hash=sha256:0cb8b8625bfdcc2d43ea4b9cdba0b39b2b7befc04f3088897031082aa16ce186\n" +
+		"virtualenv==15.2.0 --hash=sha256:e8e05d4714a1c51a2f5921e62f547fcb0f713ebbe959e0a7f585cc8bef71d11f\n" +
+		"wheel==0.31.0 --hash=sha256:9cdc8ab2cc9c3c2e2727a4b67c22881dbb0e1c503d592992594c5e131c867107\n" +
+		""
+)
+
+func getPipScript(cache stoic.Cache) (string, error) {
+	var err error
+
+	script, err := ioutil.TempFile("", "get-pip-*.py")
+	if err != nil {
+		return "", fmt.Errorf(
+			"unable to set up temp file for get-pip.py: %v", err)
+	}
+	defer func() {
+		if err != nil {
+			os.Remove(script.Name())
+		}
+		script.Close()
+	}()
+
+	if cacheReader := cache.Get(getPipCacheKey); cacheReader != nil {
+		defer cacheReader.Close()
+
+		_, err = io.Copy(script, cacheReader)
+		if err != nil {
+			return "", fmt.Errorf(
+				"unable to read get-pip.py from cache: %v", err)
+		}
+	} else {
+		resp, err := http.Get(getPipURL)
```

Comment:
Are we using the default httpClient here? If we are, I don't think it has an associated timeout, consider replacing with this:

```
timeout := time.Duration(5 * time.Second)
client := http.Client{
    Timeout: timeout,
}
client.Get(url)
```

Reply:
There are a couple of things I'd like to do with the default client. One is to add timeouts, like you suggest, the other is to add fetching of credentials from  `~/.netrc`. The reason I'm not doing anything about it at this level is that I think this should be somehow managed by the engine.

I'd say in the future the engine will either (god forbid!) override the default client, or rather offer access to a configured client through the `stoic.Stoic` interface.

I'm happy to keep that out of that PR, if you don't mind :)

Reply:
That makes perfect sense!